src/db.c

```diff
@@ -662,16 +671,19 @@ void delGenericCommand(client *c, int lazy) {
     int numdel = 0, j;
 
     for (j = 1; j < c->argc; j++) {
-        expireIfNeeded(c->db,c->argv[j]);
+        robj *key = c->argv[j];
+        /* Lookup will check if it's expired and delete it, while avoiding
+         * memory access to the expire dict by checking the hasexpire flag. */
+        if (lookupKeyWriteWithFlags(c->db, key, LOOKUP_NOTOUCH) == NULL)
+            continue;
```

Comment:
Yeah.

I suggest the db delete function should use dictUnlink on the expire dict, check the expire value (using some code refactored out of keyIsExpired), delete it and return 0 if it was logically expired. Lookup key is not needed. OK?

There's also some duplicate code between dbSyncDelete and dbAsyncDelete. I can try to factor out the common part.

Reply:
I always wanted to unify dbSyncDelete and dbSyncDelete. In fact I have done that in some fork. 
As soon as the lazy free code nullifies the value pointer, the rest of the deletion sequence can proceed normally... 
Let's do in this PR.

Reply:
Unifying dbSyncDelete and dbAsyncDelete was easy.

However, avoiding looking up each dict more than once requires a larger refactoring. The key robj is passed to `lookupKey` which calls `expireIfNeeded` calls `keyIsExpired` calls `getExpire` which looks up both dicts using the key. We'd need to pass around the dictEntry from both dicts to avoid these lookups, which is a larger change.

`expireIfNeeded` also has some other logic, besides it calls `deleteExpiredKeyAndPropagate` which calls `dbSyncDelete` or `dbAsyncDelete` so we'd get a circular dependency if `dbSyncDelete` also does what `expireIfNeeded` is doing.

The `getExpire` function is also used in many other places, but with the change that makes getExpire check the db dict and the 'hasexpire' flag before the looking up the expire dict, the change here in `delGenericCommand` is unnecessary. I'll revert it.

Looking up the db dict twice is much faster than looking up both the expire dict and the db dict.

Reply:
let's not get carried away then with complicating the code before we prove this is really beneficial for performance.
some of the changes we discussed (like unifying the delete functions) is a cleanup of it's own (prevents changes in one being forgotten in the other).

Some of the other changes we did to avoid double lookup come for free or at least without ugly changes, let's avoid the ones that are not nice unless we first prove they're really important.

> Looking up the db dict twice is much faster than looking up both the expire dict and the db dict.

did you test it, or just estimating?

Reply:
Tested.

Reply:
ok, so let's not get carried away with dirty code.

did you already publish the results / methodology somewhere?

Reply:
No... I tried to revert as much as possible (including all the changes to lookupKey) but then some tests started to fail. Not sure why yet.

Now, perhaps I shouldn't revert the changes in lookupKey, since it's ... not dirty code. You seem to prefer that the writable replica logic to be there and only call expireIfNeeded if the key exists, if I understand correctly.

Reply:
i do think we wanna do some cleanup around these areas, refactor some confusing / dangerous functions, solve some bugs, and also avoid multiple lookups if it comes for free (or make sense as a cleanup).
but the one discussed above sounds like it's going to add a lot of complexity, so if it's not causing much harm we should skip it.

note that i haven't had a look at the recent changes since i'm busy elsewhere.
i do wanna fix these 3 things (optimize access to expires dict, solve writable replica bugs, and refactor the mess around lookupKey) 7.0.
ideally we would be able to do that in two steps each that carries a title and a clear change for the release notes.
maybe we can do that by leaving the expires bit to be the second one.
but if not, one PR that does that all is ok too.
but i think i'll have more time to dig into that deeper in a couple of weeks.

Reply:
OK, now writable-replicas and lookupKey refactoring is done and I have updated this PR. I think we can resolve this thread.

I reverted this change in dbGenericDelete to avoid complicating things. It means we have an excessive lookup to the expire dict in the DEL and UNLINK commands. I think DEL and UNLINK will be faster if it can avoid looking up the expires dict in this case though, but OTOH this PR is now quite small and simple.